### ISSUE TITLE
How to do Knowledge Distillation using Tensorpack?

### ISSUE THREAD

**Issue by ErikEnglesson** [tensorpack/tensorpack#1136]
Hi,

I have implemented Knowledge Distillation[1] where I use the predictions of the "teacher" networks as labels for the "student" network during training. However, I only evaluate the teacher networks once(pre-process) on the non-augmented training set and use these as labels(even for augmented versions). 
Something like this: CE( f(x; theta_T), f( \tilde{x}; theta_S) )
where CE is the cross-entropy loss, f(x; theta_T) is the softmax output of the teacher networks on the image x and f( \tilde{x}; theta_S) is the softmax output of the student on an augmented version of x.

Now, I want to try having each augmented image use the prediction of the teacher networks for this particular image as label, i.e., CE( f(\tilde{x}, theta_T), f(\tilde{x}, theta_S) ).

How would I do this in Tensorpack? I think the main problem is how to do inference of several other networks during training in an efficient way. Do I need to write a new trainer? Where I evaluate the teacher for each batch and then run the training op? 

Any help with this would be much appreciated.

Thanks,
Erik

[1] https://arxiv.org/abs/1503.02531?context=cs

**Comment by ppwwyyxx** [comment]
```python
def build_model():
    a = tf.stop_gradient(teacher(inputs))
    b = student(inputs)
    return loss(a, b)
```

**Comment by ErikEnglesson** [comment]
Thanks for the quick reply.

You suggest to combine the graphs of the teacher networks and the student into a single new graph? Then use ChainInit or similar to load the weights for each teacher network?

I can try this, but I am a bit worried about memory. I can have several high capacity networks in the teacher and the student could have a relatively high capacity too. Is there a good way to build, load the weights and evaluate each teacher network one at a time to be able to scale to many and large teacher networks?

Thanks,
Erik

**Comment by ppwwyyxx** [comment]
You can use `tf.control_dependencies` in `build_graph` to let tensors evaluate one at a time.

**Comment by ErikEnglesson** [comment]
Ok, great. I will have a look at this. Thanks for the help!
